<launch>
    <!--
        Author: Dr. -Ing. Ahmad Kamal Nasir
        Email: dringakn@gmail.com
        Desciption: URDF Test launch file
    -->

    <arg name="use_sim_time" default="true"/>
    <env name="GAZEBO_MODEL_PATH" value="${GAZEBO_MODEL_PATH}:$(find rotors_gazebo)/models"/>
    <env name="GAZEBO_RESOURCE_PATH" value="${GAZEBO_RESOURCE_PATH}:$(find rotors_gazebo)/models"/>

    <!-- Robot model for rviz -->
    <!-- Use if the mesh files are available -->
    <!-- <arg name="model" default="$(dirname)/../urdf/test_snippets_mesh.urdf.xacro"/> -->
    <!-- Otherwise use the following -->
    <arg name="model" default="$(find gazebo_simulation)/urdf/test_snippets.urdf.xacro"/>
    <param name="robot_description" command="$(find xacro)/xacro $(arg model)" />

    <!-- Launch RViz -->
    <node type="rviz" name="rviz" pkg="rviz" args="-d $(find mav_trajectory_generation_example)/cfg/diff_drive_robot.rviz" />

    <!-- Robot transform publisher -->
    <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher"/>

    <!-- Robot joint state values -->
    <!-- Used in-case of no simulation/hardware -->
    <node pkg="joint_state_publisher" type="joint_state_publisher" name="joint_state_publisher">
        <param name="use_gui" value="false"/>
        <param name="rate" value="50"/>
    </node>

    <!-- Launch gazebo simulation and spawn model -->
    <arg name="world_name" default="$(find gazebo_simulation)/world/turtlebot_playground.world"/>
    <include file="$(find gazebo_ros)/launch/empty_world.launch">
        <arg name="world_name" value="$(arg world_name)"/>
        <arg name="paused" default="false"/>
        <arg name="use_sim_time" default="$(arg use_sim_time)"/>
        <arg name="gui" default="true"/>
        <arg name="headless" default="false"/>
        <arg name="recording" default="false"/>
        <arg name="debug" default="false"/>
    </include>

    <arg name="robot_model_name" default="my_robot"/>
    <arg name="spawn_model_args" default="-urdf -param robot_description -model $(arg robot_model_name) -x 0 -y 0 -z 0.5"/>
    <node pkg="gazebo_ros" type="spawn_model" name="spawn_model" output="screen" args="$(arg spawn_model_args)"/>

    <!-- Keyboard -->
    <include file="$(find gazebo_simulation)/launch/keyboard.launch" />

    <!--- Trajectory planner -->
    <node name="planner" pkg="mav_trajectory_generation_example" type="example_evaluate_plan" output="screen" required="true">
        <remap from="/planner/cmd_vel" to="/cmd_vel"/>
        <remap from="/planner/odom" to="/odom"/>
    </node>

    <!--- Trajectory sampler -->
    <node name="sampler" pkg="mav_trajectory_generation_ros" type="trajectory_sampler_node" output="screen">
        <remap from="path_segments_4D" to="trajectory"/>
    </node>

</launch>